### ISSUE TITLE
False UnusedVariable warning for method parameters only used in overridden implementations

### ISSUE THREAD

**Issue by mlopatkin** [google/error-prone#4409]
Consider the following example:

```java
public class App {
    private static class Base {
        protected void doStuff(String usedInDescendants) {}
    }

    private static class Descendant extends Base {
        @Override
        protected void doStuff(String actuallyUsed) {
            System.out.println(actuallyUsed);
        }
    }

    public static void main(String[] args) {
        Base b = new Descendant();

        b.doStuff("some string");
    }
}
```
**Expected result**: no warnings

**Actual result**: ErrorProne complains about `usedInDescendants` parameter of `App.Base.doStuff` being unused. However, this parameter is used by the overridden method in `Descendant` class and cannot be easily removed.

There are several workarounds, but they aren't universally applicable:
* Increase the visibility of the base class - gives the type unnecessary exposure
* Make the method abstract - then the base class cannot provide a default implementation.

Even converting the `Base` to an interface with default method still triggers the warning.

A full reproducer can be found at
https://github.com/mlopatkin/errorprone-issue-reproducers/tree/trunk/unused-variable-overloaded-method

Tested with ErrorProne 2.27.1, 2.24.1 on Java 11 and Java 21.

**Comment by cpovirk** [comment]
(Is this fixed by https://github.com/google/error-prone/pull/4433, which started off being about a regression test but seems to have developed into an actual fix?)

**Comment by mlopatkin** [comment]
Can confirm that the issue is fixed in 2.29.0. Thank you for solving it so quickly!

**Comment by mlopatkin** [comment]
As of 2.29.1, the warning still happens when the descendant is an anonymous class:
https://github.com/mlopatkin/errorprone-issue-reproducers/blob/trunk/unused-variable-overloaded-method/src/main/java/com/example/AnonymousDescendant.java
